### PR TITLE
Optimize copy_from/for_each_value

### PR DESCRIPTION
**Pull request by dsharletg**
- Use a more idiomatic loop pattern that the compiler might be more likely to understand.
- Sort strides by the strides of the buffer read, not the buffer written
- Handle the scalar special case outside `for_each_value_helper`.

### CODE REVIEW

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -1951,55 +1951,42 @@ class Buffer {
...
+        std::ptrdiff_t extent;
+        std::ptrdiff_t stride[N];
     };
 
     // Given an array of strides, and a bunch of pointers to pointers
     // (all of different types), advance the pointers using the
     // strides.
     template<typename Ptr, typename... Ptrs>
-    HALIDE_ALWAYS_INLINE static void advance_ptrs(const int *stride, Ptr *ptr, Ptrs... ptrs) {
+    HALIDE_ALWAYS_INLINE static void advance_ptrs(const std::ptrdiff_t *stride, Ptr *ptr, Ptrs... ptrs) {
```
Comment (steven-johnson) [inline src/runtime/HalideBuffer.h:1962]: Idle thought: would using references rather than pointers here enable potentially better codegen?

**Comment by steven-johnson** [comment]
LGTM. Does profiling show a measurable improvement?

File: src/runtime/HalideBuffer.h [hunk 2]
```diff
@@ -1951,55 +1951,42 @@ class Buffer {
...
-                    increment_ptrs((&ptrs)...);
+                Ptr end = ptr + t[0].extent;
+                while (ptr != end) {
+                    f(*ptr++, (*ptrs++)...);
                 }
             } else {
-                for (int i = t[0].extent; i != 0; i--) {
-                    f((*ptrs)...);
-                    advance_ptrs(t[0].stride, (&ptrs)...);
+                for (std::ptrdiff_t i = t[0].extent; i != 0; i--) {
```
Comment (abadams) [inline src/runtime/HalideBuffer.h:1981]: Why not ptr != end in this case too?

**Comment by abadams** [comment]
Makes sense to me. Just had one minor comment.

File: src/runtime/HalideBuffer.h (hunk 1 above)
Comment (dsharletg) [reply src/runtime/HalideBuffer.h:1962]: I doubt it matters, this is getting inlined. I actually would stylistically prefer references but I didn't want to change it in this PR.

**Review by abadams — APPROVED** [review]

**Pull request merged**